packages/sqs/lib/sqs/AbstractSqsConsumer.ts

```diff
@@ -226,24 +228,35 @@ export abstract class AbstractSqsConsumer<
 
         // success
         if (result.result) {
-          this.handleMessageProcessed(deserializedMessage.result, 'consumed')
+          this.handleMessageProcessed(originalMessage, 'consumed')
           return message
         }
 
         // failure
         this.handleMessageProcessed(
-          deserializedMessage.result,
+          originalMessage,
           result.error === 'retryLater' ? 'retryLater' : 'error',
         )
 
-        // in case of retryLater, if DLQ is set, we will send the message back to the queue
-        if (result.error === 'retryLater' && this.deadLetterQueueUrl) {
-          await this.sqsClient.send(
-            new SendMessageCommand({
-              QueueUrl: this.queueUrl,
-              MessageBody: message.Body,
-            }),
-          )
+        // in case of retryLater, requeue the message if maxRetryDuration is not exceeded
+        if (result.error === 'retryLater') {
+          // if timestamp is not present, defining it for the next try
+          const timestamp = this.tryToExtractTimestamp(parsedMessage) ?? new Date()
+          const lastRetryDate = new Date(timestamp.getTime() + this.maxRetryDuration * 1000)
+          if (lastRetryDate > new Date()) {
+            await this.sqsClient.send(
+              new SendMessageCommand({
+                QueueUrl: this.queueUrl,
+                MessageBody: JSON.stringify({
+                  ...originalMessage,
+                  [this.messageTimestampField]: timestamp.toISOString(),
+                }),
+              }),
+            )
+          } else {
+            await this.failProcessing(message)
+          }
+
```

Comment:
Decided to not split between barrier and handler `retryLater`, at the end in the handler is also possible to have a stuck message and we want to avoid that situation, let me know if you disagree

Reply:
makes sense!